### PR TITLE
fix: Respect ignore files, and ignore git dir when tarring sources

### PR DESCRIPTION
**Pull request by CameronDiver**
This commit pulls in the minimatch library, and uses it to remove any
globs present in .gitignore or .dockerignore files. The ignore files
themselves are also not forwarded to the tar archive (which is used in
both resin push and resin deploy).

Change-type: major
Closes: #889
Signed-off-by: Cameron Diver <cameron@resin.io>

### CODE REVIEW

File: lib/utils/compose.coffee [hunk 1]
```diff
@@ -103,24 +103,51 @@ toPosixPath = (systemPath) ->
...
-		Promise.join relPath, fs.stat(file), fs.readFile(file),
-			(filename, stats, data) ->
-				pack.entry({ name: toPosixPath(filename), size: stats.size, mode: stats.mode }, data)
+	.each (file) ->
+		# Find any files that should be ignored
+		if path.relative(path.resolve(dir), file) in ['.gitignore', '.dockerignore' ]
+			return extractIgnored(file)
+			.then (ignoreGlobs) ->
+				# add the ignored files, and also add the ignore file itself
+				ignoreGlobs.push(file)
```
Comment (pimterry) [inline lib/utils/compose.coffee]: Why do we also include the ignore file? That's not standard behaviour for Docker builds, afaict, and you might want it (maybe if you use git inside your build).

File: lib/utils/compose.coffee [hunk 2]
```diff
@@ -103,24 +103,51 @@ toPosixPath = (systemPath) ->
...
 
 	getFiles = ->
 		streamToPromise(klaw(dir))
 		.filter((item) -> not item.stats.isDirectory())
 		.map((item) -> item.path)
 
+	extractIgnored = (file) ->
+		fs.readFile(file, 'utf8').then (content) ->
+			_(content.split(/\r?\n/))
+			.filter((f) -> !_.isEmpty(f))
```
Comment (pimterry) [inline lib/utils/compose.coffee]: We should also drop any lines starting with `#`

File: lib/utils/compose.coffee [hunk 3]
```diff
@@ -103,24 +103,51 @@ toPosixPath = (systemPath) ->
...
+			return extractIgnored(file)
+			.then (ignoreGlobs) ->
+				# add the ignored files, and also add the ignore file itself
+				ignoreGlobs.push(file)
+				toIgnore = toIgnore.concat(ignoreGlobs)
+		return []
+	.then (files) ->
+		_(toIgnore).uniq().each (glob) ->
+			# Negate the glob so that we keep everything except the matches
+			files = files.filter(minimatch.filter("!#{glob}"))
```
Comment (pimterry) [inline lib/utils/compose.coffee]: This doesn't work, as rules can override each other. Taking an example from the Docker docs:

```
*.md
!README.md
```

This should mean 'ignore every .md except README.md'. With this though, we'll filter out all the md's first, so README.md will already be gone before we get to the 2nd rule.

File: lib/utils/compose.coffee [hunk 4]
```diff
@@ -103,24 +103,51 @@ toPosixPath = (systemPath) ->
...
+		# Also remove any git directory files - we don't need to worry about the .git
+		# directory itself as without any entries the directory will never be created
+		files = files.filter(minimatch.filter("!#{dir}#{path.sep}.git/**/*"))
```
Comment (pimterry) [inline lib/utils/compose.coffee]: I don't think we want to do this - you really might want to use git in your build, and this behaviour doesn't match `docker build`, which will catch people out.

File: lib/utils/compose.coffee [hunk 5]
```diff
@@ -103,24 +103,51 @@ toPosixPath = (systemPath) ->
...
 	pack = tar.pack()
 	getFiles(dir)
-	.map (file) ->
-		relPath = path.relative(path.resolve(dir), file)
-		Promise.join relPath, fs.stat(file), fs.readFile(file),
-			(filename, stats, data) ->
-				pack.entry({ name: toPosixPath(filename), size: stats.size, mode: stats.mode }, data)
+	.each (file) ->
+		# Find any files that should be ignored
+		if path.relative(path.resolve(dir), file) in ['.gitignore', '.dockerignore' ]
```
Comment (pimterry) [inline lib/utils/compose.coffee]: You're not going to like this: gitignore and dockerignore aren't the same format.

https://zzz.buzz/2018/05/23/differences-of-rules-between-gitignore-and-dockerignore/ has some details, and https://gist.github.com/wassname/b25471b0f3bb2f9ff81f might be interesting.

File: lib/utils/compose.coffee [hunk 6]
```diff
@@ -103,24 +103,51 @@ toPosixPath = (systemPath) ->
...
+		_(toIgnore).uniq().each (glob) ->
+			# Negate the glob so that we keep everything except the matches
+			files = files.filter(minimatch.filter("!#{glob}"))
+		# Also remove any git directory files - we don't need to worry about the .git
+		# directory itself as without any entries the directory will never be created
+		files = files.filter(minimatch.filter("!#{dir}#{path.sep}.git/**/*"))
+
+		Promise.map files, (file) ->
+			relPath = path.relative(path.resolve(dir), file)
+			Promise.join relPath, fs.stat(file), fs.readFile(file),
```
Comment (pimterry) [inline lib/utils/compose.coffee]: As in #943, we might need to lstat and handle symlinks here (tbc depending on the resolution of that PR)

File: lib/utils/compose.coffee (hunk 5 above)
Comment (CameronDiver) [reply lib/utils/compose.coffee]: you're right, I don't like this :/

**Comment by CameronDiver** [comment]
Closing to rebase onto v8 branch.

**Pull request closed without merging**